### ISSUE TITLE
warning: The size of tensor a (128) must match the size of tensor b (512) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([128, 3072]), vit_embeds.shape=torch.Size([512, 3072])

### ISSUE THREAD

**Issue by SuiFengLai** [hiyamdebary/EarthDial#19]
Hello, author, why did the code report it when training SAR images in the third stage of training?
warning: The size of tensor a (128) must match the size of tensor b (512) at non-singleton dimension 0, input_embeds[selected].shape=torch.Size([128, 3072]), vit _ embeds. shape = torch. size ([512, 3072]), and the loss from training is also 0.

**Comment by SuiFengLai** [comment]
Hello, author. I think the data preprocessing file src/earthdial/train/dataset.py seems to have defects in SAR data processing and UHI data processing.
Transform is defective in SAR data training mode (dataset.py:280-285);
-without T.ToTensor (), T.Normalize cannot work properly.
-eval mode (dataset.py:322-328) has T.ToTensor (), but the training mode does not.

UHI data training mode will mistakenly switch to 3-channel RGB:
-t.lambda (lambda img: img if img.mode = =' l' else img.convert ('RGB')) will convert 8 channels into 3 channels.
-MultiChannelResize like eval mode should be used to process multichannel data.

**Comment by freezingLB** [comment]
Hello, I would like to ask how I can get the training code for this model. I want to understand the training process of this model. Thank you!

**Comment by brimoon-c** [comment]
The reason of this problem is that the number of `<image>` tags in the text does not match the actual number of patch images.

Adding the following code to line 369 of the file EarthDial/src/earthdial/train/dataloader.py will resolve this issue:

raw_text = data_item["conversations"][0]["value"]
image_tag_count = raw_text.count("<image>")
        
if image_tag_count < num_image:
    missing_tags = num_image - image_tag_count
    data_item["conversations"][0]["value"] = ("<image>\n" * missing_tags) + raw_text
elif image_tag_count > num_image:
    pass
